File: payments/src/product.rs
use stripe::ProductId;

use crate::payments_v1::{product_handler_server::ProductHandler, *};

pub struct ProductClient {
    pub client: stripe::Client,
}

#[async_trait::async_trait]
impl ProductHandler for ProductClient {
    async fn get_product(
        &self,
        request: tonic::Request<GetProductRequest>,
    ) -> Result<tonic::Response<GetProductReply>, tonic::Status> {
        let request = request.into_inner();

        let parsed_id = request.id.parse::<ProductId>();
        if parsed_id.is_err() {
            let e = format!("Invalid product id: {}", request.id);
            return Err(tonic::Status::new(tonic::Code::InvalidArgument, e));
        }

        let result = stripe::Product::retrieve(&self.client, &parsed_id.unwrap(), &vec![]).await;
        if let Ok(product) = result {
            let reply = GetProductReply {
                name: product.name.unwrap_or("".to_string()),
                description: product.description.unwrap_or("".to_string()),
            };
            Ok(tonic::Response::new(reply))
        } else {
            let e = format!(
                "Internal server error - {}",
                result.unwrap_err().to_string()
            );
            Err(tonic::Status::new(tonic::Code::Internal, e))
        }
    }

    async fn create_product(
        &self,
        request: tonic::Request<CreateProductRequest>,
    ) -> Result<tonic::Response<CreateProductReply>, tonic::Status> {
        let request = request.into_inner();

        let mut product = stripe::CreateProduct::new(&request.name);
        product.description = Some(&request.description);
        product.metadata = Some(request.metadata);

        let result = stripe::Product::create(&self.client, product).await;
        if let Ok(product) = result {
            let reply = CreateProductReply {
                id: product.id.to_string(),
            };
            Ok(tonic::Response::new(reply))
        } else {
            let e = format!(
                "Internal server error - {}",
                result.unwrap_err().to_string()
            );
            Err(tonic::Status::new(tonic::Code::Internal, e))
        }
    }

    async fn delete_product(
        &self,
        request: tonic::Request<DeleteProductRequest>,
    ) -> Result<tonic::Response<DeleteProductReply>, tonic::Status> {
        let request = request.into_inner();

        let parsed_id = request.id.parse::<ProductId>();
        if parsed_id.is_err() {
            let e = format!("Invalid product id: {}", request.id);
            return Err(tonic::Status::new(tonic::Code::InvalidArgument, e));
        }

        let result = stripe::Product::delete(&self.client, &parsed_id.unwrap()).await;
        if let Ok(_) = result {
            let reply = DeleteProductReply { deleted: true };
            Ok(tonic::Response::new(reply))
        } else {
            let e = format!(
                "Internal server error - {}",
                result.unwrap_err().to_string()
            );
            Err(tonic::Status::new(tonic::Code::Internal, e))
        }
    }

    async fn update_product(
        &self,
        request: tonic::Request<UpdateProductRequest>,
    ) -> Result<tonic::Response<UpdateProductReply>, tonic::Status> {
        let request = request.into_inner();

        let parsed_id = request.id.parse::<ProductId>();
        if parsed_id.is_err() {
            let e = format!("Invalid product id: {}", request.id);
            return Err(tonic::Status::new(tonic::Code::InvalidArgument, e));
        }

        let mut product = stripe::UpdateProduct::new();
        product.name = Some(&request.name);
        product.description = Some(&request.description);

        let result = stripe::Product::update(&self.client, &parsed_id.unwrap(), product).await;
        if let Ok(_) = result {
            let reply = UpdateProductReply { updated: true };
            Ok(tonic::Response::new(reply))
        } else {
            let e = format!(
                "Internal server error - {}",
                result.unwrap_err().to_string()
            );
            Err(tonic::Status::new(tonic::Code::Internal, e))
        }
    }
}

File: stripe/src/product.go
package main

import (
	"context"
	"fmt"

	pb "github.com/lee-aaron/stripe-go/product"
	"github.com/stripe/stripe-go/v72"
	"github.com/stripe/stripe-go/v72/price"
	"github.com/stripe/stripe-go/v72/product"
	"google.golang.org/grpc"
)

type ProductServer struct {
	pb.UnimplementedProductHandlerServer
}

func (s *ProductServer) Register(server *grpc.Server, srv *ProductServer) {
	pb.RegisterProductHandlerServer(server, srv)
}

func (s *ProductServer) CreateProduct(ctx context.Context, req *pb.CreateProductRequest) (*pb.CreateProductReply, error) {
	params := &stripe.ProductParams{
		Name:        stripe.String(req.Name),
		Description: stripe.String(req.Description),
		DefaultPriceData: &stripe.ProductDefaultPriceDataParams{
			Currency: stripe.String(string(stripe.CurrencyUSD)),
			Recurring: &stripe.ProductDefaultPriceDataRecurringParams{
				Interval: stripe.String(string(stripe.PriceRecurringIntervalMonth)),
			},
			UnitAmount: stripe.Int64(req.Amount),
		},
	}

	for k, v := range req.Metadata {
		params.AddMetadata(k, v)
	}

	params.SetStripeAccount(req.StripeAccount)

	p, err := product.New(params)
	if err != nil {
		return nil, err
	}

	return &pb.CreateProductReply{
		ProdId:  p.ID,
		PriceId: p.DefaultPrice.ID,
	}, nil
}

func (s *ProductServer) UpdateProduct(ctx context.Context, req *pb.UpdateProductRequest) (*pb.UpdateProductReply, error) {
	params := &stripe.ProductParams{
		Name:        stripe.String(req.Name),
		Description: stripe.String(req.Description),
	}

	params.SetStripeAccount(req.StripeAccount)

	_, err := product.Update(req.Id, params)
	if err != nil {
		return &pb.UpdateProductReply{
			Updated: false,
		}, err
	}

	return &pb.UpdateProductReply{
		Updated: true,
	}, nil
}

func (s *ProductServer) GetProduct(ctx context.Context, req *pb.GetProductRequest) (*pb.GetProductReply, error) {
	params := &stripe.ProductParams{}
	params.SetStripeAccount(req.StripeAccount)

	p, err := product.Get(req.Id, params)
	if err != nil {
		return nil, err
	}

	return &pb.GetProductReply{
		Name:        p.Name,
		Description: p.Description,
	}, nil
}

func (s *ProductServer) DeleteProduct(ctx context.Context, req *pb.DeleteProductRequest) (*pb.DeleteProductReply, error) {
	params := &stripe.ProductParams{}
	params.SetStripeAccount(req.StripeAccount)

	p, err := product.Del(req.Id, params)
	if err != nil {
		return nil, err
	}

	return &pb.DeleteProductReply{
		Deleted: p.Deleted,
	}, nil
}

func (s *ProductServer) SearchProduct(ctx context.Context, req *pb.SearchProductRequest) (*pb.SearchProductReply, error) {
	params := &stripe.ProductSearchParams{}
	params.Limit = req.Limit
	params.Page = req.Page
	params.Query = req.Query
	params.AddExpand("data.default_price")
	params.SetStripeAccount(req.StripeAccount)
	iter := product.Search(params)

	var products []*pb.Product

	for iter.Next() {

		priceParams := &stripe.PriceSearchParams{}
		priceParams.Query = fmt.Sprintf("metadata['product_id']: '%s'", iter.Product().ID)
		priceParams.SetStripeAccount(req.StripeAccount)

		curr := price.Search(priceParams)
		if curr.Next() {
			products = append(products, &pb.Product{
				Id:          iter.Product().ID,
				Name:        iter.Product().Name,
				Description: iter.Product().Description,
				Metadata:    iter.Product().Metadata,
				Amount:      curr.Price().UnitAmount,
			})
		}
	}

	return &pb.SearchProductReply{
		Products: products,
	}, nil
}
